language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -276,146 +351,92 @@ computeCek con t@(Var _ varName)   = do
-    | otherwise                      =
-        throwingWithCause _MachineError NonPrimitiveInstantiationMachineError $ Just fun
-
-{- Note [Saved mapping example]
-Consider a polymorphic built-in function @id@, whose type signature on the Plutus side is
-
-    id : all a. a -> a
-
-Notation:
-
-- the variable environment is denoted as @{ <var> :-> (<env>, <value>), ... }@
-- the context is denoted as a Haskell list of 'Frame's
-- 'computeCek' is denoted as @(|>)@
-- 'returnCek' is denoted as @(<|)@
-
-When evaluating
+    => Context uni -> Type TyName uni ExMemory -> CekVal uni -> CekM uni (Plain Term uni)
+instantiateEvaluate ctx _ (VTyAbs _ _ _ body env) = withEnv env $ computeCek ctx body -- FIXME: env?
+instantiateEvaluate ctx ty (VBuiltin ex argEnv bn count tyargs args) =
+    returnCek ctx $ VBuiltin ex argEnv bn (count-1) (ty:tyargs) args
+    -- We should really check that the signature expects a type here, not a term.
+instantiateEvaluate _ _ val =
+        throwingWithCause _MachineError NonPolymorphicInstantiationMachineError $ Just val
 
-    id {integer -> integer} ((\(i : integer) (j : integer) -> i) 1) 0
-
-We encounter the following state:
-
-    { i :-> ({}, 1) }
-    [ FrameApplyFun {} (id {integer -> integer})
-    , FrameApplyArg {} 0
-    ] |> \(j : integer) -> i
-
-and transition it into
-
-    { arg :-> ({ i :-> ({}, 1) }, \(j : integer) -> i) }
-    [ FrameApplyArg {} 0
-    ] <| id {integer -> integer} arg
-
-i.e. if the argument is not a constant, then we create a new variable, save the old environment in
-the closure of that variable and apply the function to the variable. This allows to restore the old
-environment @{ i :-> ({}, 1) }@ latter when we start evaluating @arg 0@, which expands to
-
-    (\(j : integer) -> i)) 0
-
-which evaluates to @1@ in the old environment.
--}
-
--- See Note [Saved mapping example].
--- See https://github.com/input-output-hk/plutus/issues/1882 for discussion
--- | If an argument to a built-in function is a constant, then feed it directly to the continuation
--- that handles the argument and invoke the continuation in the caller's environment.
--- Otherwise create a fresh variable, save the environment of the argument in a closure, feed
--- the created variable to the continuation and invoke the continuation in the caller's environment
--- extended with a mapping from the created variable to the closure (i.e. original argument +
--- its environment). The "otherwise" is only supposed to happen when handling an argument to a
--- polymorphic built-in function.
-withScopedArgIn
-    :: VarEnv uni                            -- ^ The caller's envorinment.
-    -> VarEnv uni                            -- ^ The argument's environment.
-    -> WithMemory Value uni                  -- ^ The argument.
-    -> (WithMemory Value uni -> CekM uni a)  -- ^ A continuation handling the argument.
-    -> CekM uni a
-withScopedArgIn funVarEnv _         arg@Constant{} k = withVarEnv funVarEnv $ k arg
```

Comment:
If we're getting rid of this stuff we'll probably want to update any notes about polymorphic builtins.